compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+- LZ77 sequences - we first process the match_length by subtracting min_match_length and match_offset
+by subtracting 1. Then we concatenate the literal counts and the processed match lengths and offsets
+and pass them all into an Elias Delta code.
+
+
+Current limitations: 
+1. During compression we use a naive match finding whose
+   memory usage accumulates over time. We could implement the hash-chain algorithm 
+   described at https://glinscott.github.io/lz/index.html#toc4.2.2.
+2. During decompression we keep the entire past output buffer in memory. We
+   could implement window-based scheme and implement shifting the window to
+   reduce memory usage.
+3. We implement greedy parsing that takes the longest match at a given position.
+   Instead we could implement more optimal parsing which tries to skip some
+   bytes as literals to find a longer match in the upcoming bytes.
+4. We do not perform any special handling to limit complexity for edge cases like
+   all very long repeats which have a high complexity of finding the longest match.
+
+Given the above, note that you could still reset the encoder in order after a
+few blocks to limit memory usage.
+
+Benchmarks on a few files from https://github.com/nemequ/squash-corpus and 
+https://corpus.canterbury.ac.nz/descriptions/#cantrbry (plus a few handmade).
+
+All sizes are in bytes.
+
+| File                                | raw size | lz77 size (this) | gzip size |
+|-------------------------------------|----------|------------------|-----------|
+| bootstrap-3.3.6.min.css             |121260    |23526             |19747      |
+| eff.html                            |41684     |10935             |9670       |
+| zlib.wasm                           |86408     |43471             |37448      |
+| jquery-2.1.4.min.js                 |84345     |34447             |29569      |
+| random.bin (random bytes)           |1000000   |1004970           |1000328    |
+| repeat_As.txt                       |1000000   |208               |1004       |
+| kennedy.xls                         |1029744   |330252            |204004     |
+| alice29.txt                         |152089    |63376             |54416      |
+
+"""
+
+import argparse
+from dataclasses import dataclass
+import os
+import tempfile
+from typing import List, Tuple
+from compressors.elias_delta_uint_coder import EliasDeltaUintDecoder, EliasDeltaUintEncoder
+from compressors.huffman_coder import HuffmanDecoder, HuffmanEncoder
+from core.data_block import DataBlock
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from core.data_stream import Uint8FileDataStream
+from core.encoded_stream import EncodedBlockReader, EncodedBlockWriter
+from core.prob_dist import ProbabilityDist
+from utils.bitarray_utils import BitArray, bitarray_to_uint, uint_to_bitarray
+from utils.test_utils import (
+    create_random_binary_file,
+    try_file_lossless_compression,
+    try_lossless_compression,
+)
+
+ENCODED_BLOCK_SIZE_HEADER_BITS = (
```

Comment:
@shubhamchandak94  -> ok for now, but maybe add this in a 

```
@dataclass
class LZ77Params:
          ENCODED_BLOCK_SIZE ... 
          

```

And use this by default in the encoder/decoder.

Reply:
Can you please add an issue to fix this?

Reply:
Actually nvm, you have these parameters as separate inputs.